notebooks/02_model/hypertune_spark_deep_dive.ipynb

```diff
@@ -0,0 +1,983 @@
+    "    }\n",
+    "    \n",
+    "    als = ALS(\n",
+    "        rank=rank,\n",
+    "        maxIter=15,\n",
+    "        implicitPrefs=False,\n",
+    "        alpha=0.1,\n",
+    "        regParam=reg,\n",
+    "        coldStartStrategy='drop',\n",
+    "        nonnegative=False,\n",
+    "        seed=0,\n",
+    "        **header\n",
+    "    )\n",
+    "    \n",
+    "    model = als.fit(train)\n",
+    "    \n",
+    "    users = train.select(col_user).distinct()\n",
+    "    items = train.select(col_item).distinct()\n",
+    "    user_item = users.crossJoin(items)\n",
+    "    dfs_pred = model.transform(user_item)\n",
+    "\n",
+    "    # Remove seen items.\n",
+    "    dfs_pred_exclude_train = dfs_pred.alias(\"pred\").join(\n",
+    "        train.alias(\"train\"),\n",
+    "        (dfs_pred[col_user] == train[col_user]) & (dfs_pred[col_item] == train[col_item]),\n",
+    "        how='outer'\n",
+    "    )\n",
+    "\n",
+    "    top_all = dfs_pred_exclude_train.filter(dfs_pred_exclude_train[\"train.Rating\"].isNull()) \\\n",
+    "        .select('pred.' + col_user, 'pred.' + col_item, 'pred.' + \"prediction\")\n",
+    "    \n",
+    "    top_all.cache().count()\n",
+    "\n",
+    "    rank_eval = SparkRankingEvaluation(\n",
+    "        valid, \n",
+    "        top_all, \n",
+    "        k=k, \n",
+    "        col_user=col_user, \n",
+    "        col_item=col_item, \n",
+    "        col_rating=\"Rating\", \n",
+    "        col_prediction=\"prediction\", \n",
+    "        relevancy_method=relevancy_method\n",
+    "    )\n",
+    "    \n",
+    "    precision = rank_eval.precision_at_k()\n",
+    "    \n",
+    "    # Return the objective function result.\n",
+    "    return {\n",
+    "        'loss': -precision,\n",
+    "        'status': STATUS_OK,\n",
+    "        'eval_time': time.time()\n",
+    "    }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Hyperparameter tuning with `hyperopt` sampling methods"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Though `hyperopt` works well in a single node machine, its features (e.g., `Trials` module) do not support Spark environment. A good practice is to use `hyperopt` for sampling parameter values from the defined sampling space, and then parallelize the model training onto Spark cluster with the sampled parameter combinations.\n",
+    "\n",
+    "The downside of this method is that the intelligent searching algorithm (i.e., TPE) of `hyperopt` cannot be used. The approach introduced here is therefore equivalent to random search."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Sample the parameters used for model building from the pre-defined space. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import hyperopt.pyll.stochastic\n",
```

Comment:
In other notebooks we put all the imports in the top cell

Reply:
We changed the deep dive notebooks structure a little bit. E.g., in the reco algorithm notebooks, we put theoretical introduction at the beginning suggested by @wutaomsft.